internal/api/token_oidc.go

```diff
@@ -0,0 +1,181 @@
+
+		allowed := false
+		for _, allowedIssuer := range config.External.AllowedIdTokenIssuers {
+			if p.Issuer == allowedIssuer {
+				allowed = true
+				providerType = allowedIssuer
+				acceptableClientIDs = []string{p.ClientID}
+				issuer = allowedIssuer
+				break
+			}
+		}
+
+		if !allowed {
+			return nil, "", nil, badRequestError(fmt.Sprintf("Custom OIDC provider %q not allowed", p.Issuer))
+		}
+	}
+
+	if !enabled {
+		return nil, "", nil, badRequestError(fmt.Sprintf("Provider (issuer %q) is not enabled", issuer))
+	}
+
+	oidcProvider, err := oidc.NewProvider(ctx, issuer)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	return oidcProvider, providerType, acceptableClientIDs, nil
+}
+
+// IdTokenGrant implements the id_token grant type flow
+func (a *API) IdTokenGrant(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	db := a.db.WithContext(ctx)
+	config := a.config
+
+	params := &IdTokenGrantParams{}
+
+	body, err := getBodyBytes(r)
+	if err != nil {
+		return badRequestError("Could not read body").WithInternalError(err)
+	}
+
+	if err := json.Unmarshal(body, params); err != nil {
+		return badRequestError("Could not read id token grant params: %v", err)
+	}
+
+	if params.IdToken == "" {
+		return oauthError("invalid request", "id_token required")
+	}
+
+	if params.Provider == "" && (params.ClientID == "" || params.Issuer == "") {
+		return oauthError("invalid request", "provider or client_id and issuer required")
+	}
+
+	oidcProvider, providerType, acceptableClientIDs, err := params.getProvider(ctx, config, r)
+	if err != nil {
+		return err
+	}
+
+	idToken, userData, err := provider.ParseIDToken(ctx, oidcProvider, nil, params.IdToken, "")
+	if err != nil {
+		return oauthError("invalid request", "Bad ID token").WithInternalError(err)
+	}
+
+	if idToken.Subject == "" {
+		return oauthError("invalid request", "Missing sub claim in id_token")
+	}
+
+	correctAudience := false
+	for _, clientID := range acceptableClientIDs {
+		for _, aud := range idToken.Audience {
+			if aud == clientID {
+				correctAudience = true
+				break
+			}
+		}
+
+		if correctAudience {
+			break
+		}
+	}
+
+	if !correctAudience {
+		return oauthError("invalid request", "Unacceptable audience in id_token")
+	}
+
+	tokenHasNonce := idToken.Nonce != ""
+	paramsHasNonce := params.Nonce != ""
+
+	if tokenHasNonce != paramsHasNonce {
+		return oauthError("invalid request", "Passed nonce and nonce in id_token should either both exist or not.")
+	} else if tokenHasNonce && paramsHasNonce {
+		// verify nonce to mitigate replay attacks
+		hash := fmt.Sprintf("%x", sha256.Sum256([]byte(params.Nonce)))
+		if hash != idToken.Nonce {
+			return oauthError("invalid nonce", "Nonces mismatch")
+		}
+	}
+
+	if err := db.Transaction(func(tx *storage.Connection) error {
+		if _, terr := a.createAccountFromExternalIdentity(tx, r, userData, providerType); terr != nil {
```

Comment:
This issues a redirect I believe, and in this case it should just return the access and refresh tokens as a body response. I will address this.